host/settings/pin/pin.go

```diff
@@ -0,0 +1,296 @@
+package pin
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/shopspring/decimal"
+	"go.sia.tech/core/types"
+	"go.sia.tech/hostd/host/settings"
+	"go.uber.org/zap"
+)
+
+type (
+	// A Pin is a pinned price in an external currency.
+	Pin struct {
+		Pinned bool    `json:"pinned"`
+		Value  float64 `json:"value"`
+	}
+
+	// PinnedSettings contains the settings that can be optionally
+	// pinned to an external currency. This uses an external explorer
+	// to retrieve the current exchange rate.
+	PinnedSettings struct {
+		// Currency is the external three letter currency code. If empty,
+		// pinning is disabled. If the explorer does not support the
+		// currency an error is returned.
+		Currency string `json:"currency"`
+
+		// Threshold is a percentage from 0 to 1 that determines when the
+		// host's settings are updated based on the current exchange rate.
+		Threshold float64 `json:"threshold"`
+
+		// Storage, Ingress, and Egress are the pinned prices in the
+		// external currency.
+		Storage Pin `json:"storage"`
+		Ingress Pin `json:"ingress"`
+		Egress  Pin `json:"egress"`
+
+		// MaxCollateral is the maximum collateral that the host will
+		// accept in the external currency.
+		MaxCollateral Pin `json:"maxCollateral"`
+	}
+
+	// A SettingsManager updates and retrieves the host's settings.
+	SettingsManager interface {
+		Settings() settings.Settings
+		UpdateSettings(settings.Settings) error
+	}
+
+	// A Store stores and retrieves pinned settings.
+	Store interface {
+		PinnedSettings() (PinnedSettings, error)
+		UpdatePinnedSettings(PinnedSettings) error
+	}
+
+	// An ExchangeRateRetriever retrieves the current exchange rate from
+	// an external source.
+	ExchangeRateRetriever interface {
+		SiacoinExchangeRate(ctx context.Context, currency string) (float64, error)
+	}
+
+	// A Manager manages the host's pinned settings and updates the host's
+	// settings based on the current exchange rate.
+	Manager struct {
+		log      *zap.Logger
+		store    Store
+		explorer ExchangeRateRetriever
+		sm       SettingsManager
+
+		frequency  time.Duration
+		rateWindow time.Duration
+
+		mu       sync.Mutex
+		rates    []decimal.Decimal
+		lastRate decimal.Decimal
+		settings PinnedSettings // in-memory cache of pinned settings
```

Comment:
Oh ok, so I guess because there's an in-memory cache the getter doesn't need a context. I'd still define it on the interface for potential implementations that don't have an in-memory cache 🤷‍♂️  but I see why you didn't add one now.